src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs

```diff
@@ -1774,19 +1774,31 @@ private Symbol GetDeclaredMember(NamespaceOrTypeSymbol container, TextSpan decla
                     }
                 }
 
-                foreach (var loc in symbol.Locations)
+                // Get the first location, and see if it's empty.  If it's not empty, then this was a member that parsed
+                // reasonably (i.e. it had a location for the name).  As such, all of its parts in all files must also
+                // have parsed in a similar fashion (otherwise they could not have stitched together into the same
+                // namespace/type).  As such, this symbol is the best one to return as long as it's defined in this
+                // tree/span.
+                //
+                // If, however, the location name is empty, then this was a bogus symbol created in an error recovery
+                // scenario.  Remember this symbol for the end in case we aren't able to find anything better to return.
+                var firstLocation = symbol.TryGetFirstLocation();
+                if (firstLocation != null && !firstLocation.SourceSpan.IsEmpty)
                 {
-                    if (loc.IsInSource && loc.SourceTree == this.SyntaxTree && declarationSpan.Contains(loc.SourceSpan))
+                    if (symbol.IsDefinedInSourceTree(this.SyntaxTree, declarationSpan))
+                        return symbol;
+                }
+                else
```

Comment:
💭 It seems like this `else` is invalid (would prevent certain cases from locating a symbol.